crates/core/src/host/module_host.rs

```diff
@@ -145,151 +144,126 @@ impl From<DatabaseTableUpdate> for TableUpdateJson {
-    pub op_type: OpType,
-    pub row: &'a ProductValue,
-}
-
-impl<'a> TableOpRef<'a> {
-    #[inline]
-    pub fn new(op_type: OpType, row: &'a ProductValue) -> Self {
-        Self { op_type, row }
-    }
-
-    #[inline]
-    pub fn insert(row: &'a ProductValue) -> Self {
-        Self::new(OpType::Insert, row)
+impl From<&UpdatesRV<'_>> for Vec<TableRowOperationJson> {
+    fn from(updates: &UpdatesRV<'_>) -> Self {
+        updates
+            .iter()
+            .map(|(op, row)| rv_to_tro_json(row.clone(), op))
+            .collect()
     }
+}
 
-    #[inline]
-    pub fn delete(row: &'a ProductValue) -> Self {
-        Self::new(OpType::Delete, row)
+impl From<&UpdatesRV<'_>> for Vec<ProductValue> {
+    fn from(updates: &UpdatesRV<'_>) -> Self {
+        updates
+            .iter()
+            .map(|(_, row)| row.clone().into_product_value())
+            .collect()
     }
 }
 
-impl From<OpType> for OperationType {
-    fn from(op: OpType) -> Self {
-        match op {
-            OpType::Delete => OperationType::Delete,
-            OpType::Insert => OperationType::Insert,
-        }
-    }
+#[repr(u8)]
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub enum OpType {
+    Delete = 0,
+    Insert = 1,
 }
 
-impl From<TableOpRef<'_>> for TableRowOperation {
-    fn from(top: TableOpRef<'_>) -> Self {
-        let row = to_vec(top.row).unwrap();
-        let op: OperationType = top.op_type.into();
-        Self { op: op.into(), row }
+impl OpType {
+    /// Converts the type to its JSON representation.
+    fn as_json_str(self) -> &'static str {
+        match self {
+            Self::Delete => "delete",
+            Self::Insert => "insert",
+        }
     }
 }
 
-impl From<TableOpRef<'_>> for TableRowOperationJson {
-    fn from(top: TableOpRef<'_>) -> Self {
-        TableOp::from(top).into()
+impl From<OpType> for OperationType {
+    fn from(op_ty: OpType) -> Self {
+        match op_ty {
+            OpType::Delete => Self::Delete,
+            OpType::Insert => Self::Insert,
+        }
     }
 }
 
-impl From<TableOpRef<'_>> for TableOp {
-    fn from(top: TableOpRef<'_>) -> Self {
-        let row = top.row.clone();
-        let op_type = top.op_type;
-        Self { op_type, row }
+impl From<OpType> for i32 {
+    fn from(op_ty: OpType) -> Self {
+        OperationType::from(op_ty) as i32
     }
 }
 
-#[derive(Debug, Clone, PartialEq, Eq)]
-pub struct TableOp {
-    pub op_type: OpType,
-    pub row: ProductValue,
+/// Annotate `row` with `op` as a `TableRowOperationJson`.
+pub(crate) fn rv_to_tro_json(row: RelValue<'_>, op: OpType) -> TableRowOperationJson {
+    pv_to_tro_json(row.into_product_value(), op)
 }
 
-impl TableOp {
-    #[inline]
-    pub fn new(op_type: OpType, row: ProductValue) -> Self {
-        Self { op_type, row }
-    }
+/// Annotate `row` BSATN-encoded with `op` as a `TableRowOperation`.
+pub(crate) fn rv_to_tro_binary(scratch: &mut Vec<u8>, row: &RelValue<'_>, op: OpType) -> TableRowOperation {
```

Comment:
I hate these abbreviations with a deep passion. Please rename this and `rv_to_tro_json` to `rel_value_to_table_row_operation_{format}`.